### PR TITLE
Add Tuple input and token support to all-gather and reduce-scatter.

### PR DESCRIPTION
**Pull request by hjm-aws**

### CODE REVIEW

**Comment by gbaned** [comment]
@hjm-aws Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
@hjm-aws Can you please sign CLA. Thank you!

**Comment by hjm-aws** [comment]
> @hjm-aws Can you please sign CLA. Thank you!
Done. Thanks!

**Comment by gbaned** [comment]
@hjm-aws It still shows CLA is pending, can you please sign CLA. Thank you!

File: tensorflow/compiler/xla/client/xla_builder.cc [hunk 1]
```diff
@@ -2752,10 +2752,30 @@ XlaOp XlaBuilder::AllGatherImpl(const XlaOp operand,
     HloInstructionProto instr;
     TF_ASSIGN_OR_RETURN(const Shape* operand_shape, GetShapePtr(operand));
 
-    TF_ASSIGN_OR_RETURN(
-        Shape inferred_shape,
-        ShapeInference::InferAllGatherShape({operand_shape},
-                                            all_gather_dimension, shard_count));
+    std::vector<const Shape*> operand_shapes;
```
Comment (cheshire) [inline tensorflow/compiler/xla/client/xla_builder.cc:2755]: Could you also update documentation of semantics in operation_semantics.md?

**Review by cheshire — CHANGES_REQUESTED** [review]
Thanks a lot! Overall this makes sense, let me also check internally.

QQ: this is not decomposable, right? E.g. changes inside XLA could not be split from builder changes?

**Comment by cheshire** [comment]
Thanks, overall this looks like a very good change! Added @Kariddi and @blakehechtman for clarifications.

File: tensorflow/compiler/xla/service/hlo_verifier_test.cc [hunk 2]
```diff
@@ -2348,6 +2347,57 @@ TEST_F(HloVerifierTest, ReduceScatterNonUniformGroups) {
               HasSubstr("Replica groups expected to be of uniform size"));
 }
 
+TEST_F(HloVerifierTest, ReduceScatterTwoTokens) {
```
Comment (jurahul) [inline tensorflow/compiler/xla/service/hlo_verifier_test.cc:2350]: I think there is some confusion here between tokens and tuples. I think the intent is to add tuple support, so we should remove any mention of tokens in test names or in the change description.

File: tensorflow/compiler/xla/client/xla_builder.cc [hunk 3]
```diff
@@ -2752,10 +2752,30 @@ XlaOp XlaBuilder::AllGatherImpl(const XlaOp operand,
...
+    std::vector<XlaOp> operands;
+    if (operand_shape->IsTuple()) {
```
Comment (jurahul) [inline tensorflow/compiler/xla/client/xla_builder.cc:2757]: This looks like copy/pasted from all-reduce. Is it possible to factor it out into a helper function?

File: tensorflow/compiler/xla/client/xla_builder_test.cc [hunk 4]
```diff
@@ -407,6 +407,24 @@ TEST_F(XlaBuilderTest, AllGatherR2) {
       ShapeUtil::Equal(root->shape(), ShapeUtil::MakeShape(F32, {4, 64})));
 }
 
+TEST_F(XlaBuilderTest, AllGatherWithToken) {
```
Comment (jurahul) [inline tensorflow/compiler/xla/client/xla_builder_test.cc:410]: AllGatherWithTuple?

File: tensorflow/compiler/xla/client/xla_builder_test.cc [hunk 5]
```diff
@@ -433,6 +451,38 @@ TEST_F(XlaBuilderTest, ReduceScatter) {
       ShapeUtil::Equal(root->shape(), ShapeUtil::MakeShape(F32, {4, 8})));
 }
 
+TEST_F(XlaBuilderTest, ReduceScatterWithToken) {
```
Comment (jurahul) [inline tensorflow/compiler/xla/client/xla_builder_test.cc:454]: ReduceScatterWithTuple?

File: tensorflow/compiler/xla/service/hlo_verifier.cc [hunk 6]
```diff
@@ -406,10 +406,20 @@ static Status CheckCommonAllGatherInvariants(HloInstruction* hlo,
                                                ag->use_global_device_ids()));
   TF_RETURN_IF_ERROR(CheckReplicaGroups(ag, group_mode));
   TF_RET_CHECK(ag->all_gather_dimension() >= 0);
+  TF_RET_CHECK(ag->operand_count() >= 1);
 
   int64_t shard_count;
+  // There can be one token in the input Tuple. The token is a scalar or
```
Comment (jurahul) [inline tensorflow/compiler/xla/service/hlo_verifier.cc:412]: This is confusing IMO. Can we get clarification as to what is intent of the token input to all-gather? Also treating a scalar as a token is confusing.

**Review by jurahul — CHANGES_REQUESTED** [review]

**Comment by jurahul** [comment]
Overall it seems this is attempting to add tuple support for all-gather and reduce-scatter as well as add a optional dummy token input to the all-gather, the purpose of which is unclear.

I think we should split this into 2 PRs, one for tuple support and discuss support for token types in all-gather separately before adding it.

**Comment by gbaned** [comment]
@hjm-aws Can you please check @jurahul's comments and keep us posted ? Thank you!

**Comment by gbaned** [comment]
@hjm-aws Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
@hjm-aws Any update on this PR? Please. Thank you!

**Comment by jeffhataws** [comment]
> Overall it seems this is attempting to add tuple support for all-gather and reduce-scatter as well as add a optional dummy token input to the all-gather, the purpose of which is unclear.
> 
> I think we should split this into 2 PRs, one for tuple support and discuss support for token types in all-gather separately before adding it.

@gbaned @jurahul , the revived PR on openxla is https://github.com/openxla/xla/pull/5740 .

**Pull request closed without merging**